src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let newState = { ...state };
+  const stateHistory = [];
+
+  for (let i = 0; i < actions.length; i++) {
+    switch (actions[i].type) {
+      case 'addProperties':
+        newState = { ...newState, ...actions[i].extraData };
+        stateHistory.push({ ...newState });
+        break;
+
+      case 'removeProperties':
+        for (let k = 0; k < actions[i].keysToRemove.length; k++) {
+          if (newState.hasOwnProperty(actions[i].keysToRemove[k])) {
+            delete newState[actions[i].keysToRemove[k]];
```

Comment:
The 'hasOwnProperty' check is unnecessary here since the 'delete' operator will not throw an error if the property does not exist. You can safely delete the property without checking for its existence.